### PR TITLE
Model

### LINKED ISSUE #20: API to data source
For all data sources connect via API to build database from scratch

### PR DESCRIPTION
**Pull request by owenmwilliams**
Closes #20 for parks, population, and location data for counties; postgres table built for population data

Need new issue for:
 - Updating population API to test for latest available
 - Automated postgres build for parks, location data
 - API for airport data from FAA, DOT, etc
 - Calculating distance from polygon to point in postgis for nearest airport / nearest park calculations

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /dat
 *.csv
 *.txt
+!requirements.txt
```
Comment (grw300) [inline .gitignore:4]: Was this required? I'm surprised.

File: models/dtl/api/dtl_census.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import requests
+import json
+import pandas as pd
+
+# Returns county population data from Census API - static from 2019; UPDATE to be dynamic search for latest
+
+def census_get():
+    response = requests.get('https://api.census.gov/data/2019/pep/population?get=NAME,POP&for=county:*&key=732b0a009f519f9728b9e054d659e8ecf2d9e6fc')
+    x = response.json()
+    return pd.read_json(json.dumps(x))
```
Comment (grw300) [inline models/dtl/api/dtl_census.py]: Add a newline at the end (that's what the Θ symbol on this line is asking for).

File: models/dtl/api/dtl_loc.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import json
+import pandas as pd
+
+# Returns geometry rings for counties from ArcGIS API; returns 2000 rows - UPDATE to return all counties, calc (lat, long) Centroids
+
+def shp_get():
+    response = requests.get('https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Counties_Generalized/FeatureServer/0/query?where=1%3D1&outFields=NAME,STATE_NAME,STATE_FIPS,CNTY_FIPS,HSE_UNITS,SQMI,OBJECTID&outSR=4326&f=json')
+    d = response.json()
+    loc_data = pd.json_normalize(d['features'])
+    return loc_data
```
Comment (grw300) [inline models/dtl/api/dtl_loc.py]: Newline.

File: models/dtl/api/dtl_parks.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import requests, json, pandas as pd
+
+# Returns parks (lat, long) from National Park Service API
+
+def parks_get():
+    endpoint = "https://developer.nps.gov/api/v1/parks"
+    HEADERS = {"X-Api-Key": "Y1YtARnlUdfs8BZ6B74c0jSgQ0M8dZ2uI6cplvsN"}
```
Comment (grw300) [inline models/dtl/api/dtl_parks.py]: In general it's a bad idea to check in private credentials - it's very difficult to remove them from git completely. We can discuss techniques for passing in such values through environment variables.

File: models/dtl/api/dtl_loc.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import requests
+import json
+import pandas as pd
+
+# Returns geometry rings for counties from ArcGIS API; returns 2000 rows - UPDATE to return all counties, calc (lat, long) Centroids
+
+def shp_get():
+    response = requests.get('https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Counties_Generalized/FeatureServer/0/query?where=1%3D1&outFields=NAME,STATE_NAME,STATE_FIPS,CNTY_FIPS,HSE_UNITS,SQMI,OBJECTID&outSR=4326&f=json')
```
Comment (grw300) [inline models/dtl/api/dtl_loc.py:8]: To deal with the long structure of URLs such as this, the `requests` module has a way to pass query paramenters (everything after the `query?`) [with code](https://requests.readthedocs.io/en/master/user/quickstart/#passing-parameters-in-urls). Try to convert to that.

I'm also concerned about the pseudo-random string after the base url. Is that an API key? If so, it probably shouldn't be checked in, and this URL should be [joined together](https://docs.python.org/3/library/urllib.parse.html?highlight=urlparse#urllib.parse.urljoin).

File: models/dtl/jprint.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import json
+
+def jprint(obj):
+    # create a formatted string of the Python JSON object
+    text = json.dumps(obj, sort_keys=True, indent=4)
+    print(text)
```
Comment (grw300) [inline models/dtl/jprint.py]: Newline.

File: models/est/db/cur.py [hunk 7]
```diff
@@ -3,4 +3,4 @@
 def con_cur():
     con = psycopg2.connect(database='owenwilliams', host="localhost", port="5434")
     cur = con.cursor()
-    return cur
\ No newline at end of file
+    return cur, con
```
Comment (grw300) [inline models/est/db/cur.py]: Newline.

File: models/dtl/wrt/pop_table.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+                state varchar(20),
+                county varchar(20))
+            """).format(sql.Identifier(a)))
+        cur.copy_from(buffer, a, sep=";")
+    except (Exception, psycopg2.DatabaseError) as error:
+            print("Error: %s" % error)
+            con.rollback()
+            cur.close()
+    con.commit()
+    cur.close()
```
Comment (grw300) [inline models/dtl/wrt/pop_table.py]: Newline.

**Review by grw300 — CHANGES_REQUESTED** [review]
We definitely want to discuss strategies for keeping your API key out of code before we merge this in.

File: models/dtl/api/dtl_census.py (hunk 2 above)
Comment (owenmwilliams) [reply models/dtl/api/dtl_census.py]: Why are they always looking for new lines? Is it so like if they string it all together it's readable?

File: .gitignore (hunk 1 above)
Comment (owenmwilliams) [reply .gitignore:4]: I mean...I guess so. It wasn't pushing requirements if I just had *.txt

File: models/dtl/api/dtl_loc.py (hunk 3 above)
Comment (owenmwilliams) [reply models/dtl/api/dtl_loc.py]: Added.

File: models/dtl/api/dtl_parks.py (hunk 4 above)
Comment (owenmwilliams) [reply models/dtl/api/dtl_parks.py]: Pulled out to a separate text file for now. Can we chat re: env variables at some point this weekend?

To add issue:
 - API Keys: pass private keys through environment variables

File: models/dtl/api/dtl_loc.py (hunk 5 above)
Comment (owenmwilliams) [reply models/dtl/api/dtl_loc.py:8]: Yeah - I haven't unpacked the ArcGIS API yet; neither the census one for that matter. But it does WORK; gimme the WIN!

Will open an issue:
 - Update ArcGIS API: pass params, remove limit, figure out what pseudo-random string is

File: models/dtl/jprint.py (hunk 6 above)
Comment (owenmwilliams) [reply models/dtl/jprint.py]: Added.

File: models/est/db/cur.py (hunk 7 above)
Comment (owenmwilliams) [reply models/est/db/cur.py]: Added.

File: models/dtl/wrt/pop_table.py (hunk 8 above)
Comment (owenmwilliams) [reply models/dtl/wrt/pop_table.py]: Added.

File: models/dtl/api/dtl_census.py (hunk 2 above)
Comment (grw300) [reply models/dtl/api/dtl_census.py]: It has to do with the technical definition of a text file and how diffing utilities deal with the ambiguity of files that don't end with newlines. Good write up [here](https://unix.stackexchange.com/a/18789).

File: .gitignore (hunk 1 above)
Comment (grw300) [reply .gitignore:4]: Oh, yea. Duh.

File: models/dtl/api/dtl_parks.py (hunk 4 above)
Comment (grw300) [reply models/dtl/api/dtl_parks.py]: Covered by #29

File: models/dtl/api/dtl_loc.py (hunk 5 above)
Comment (grw300) [reply models/dtl/api/dtl_loc.py:8]: Covered by #31

**Review by grw300 — APPROVED** [review]

**Pull request merged**